tests/wpt/tests/tools/wptrunner/wptrunner/browsers/servodriver.py

```diff
@@ -126,9 +129,50 @@ def make_command(self):
         return [self.webdriver_binary, f"--webdriver={self.port}"] + self.webdriver_args
 
     def cleanup(self):
-        super().cleanup()
         os.remove(self.hosts_path)
 
+    def is_alive(self):
+        # This is broken. It is always True.
+        if not super().is_alive():
+            return False
+        try:
+            requests.get(f"http://{self.host}:{self.port}/status", timeout=3)    
+        except requests.exceptions.Timeout:
+            # FIXME: This indicates a hanged browser. Reasons need to be investigated further.
+            self.logger.debug("Servo webdriver status request timed out.")
+            return True
+        except Exception as e:
+            self.logger.debug(f"Servo has shut down normally. {e}")
+            return False
+
+        return True
+
+    def stop(self, force=False):
+        retry_cnt = 0
+        while self.is_alive():
+            self.logger.info("Trying to shut down gracefully by extension command")
+            try:
+                requests.delete(
+                    f"http://{self.host}:{self.port}/session/dummy-session-id/servo/shutdown",
+                    timeout=3
+                )
+            except requests.exceptions.ConnectionError:
+                self.logger.debug("Browser already shut down (connection refused)")
+                break
+            except requests.exceptions.RequestException as e:
+                self.logger.debug(f"Request exception: {e}")
+                break
+            except requests.exceptions.Timeout:
+                self.logger.debug("Request timed out")
+                break
+                
+            retry_cnt += 1
+            if retry_cnt >= self.shutdown_retry:
+                self.logger.debug("Max retry exceeded.")
+                break
```

Comment:
Maybe this should also be a `warn` as well?

```suggestion
            retry_count += 1
            if retry_count >= self.shutdown_retry:
                self.logger.debug("Max number of retries exceeded. Continuing without shutting down.")
                break
```